### PR TITLE
Arithmetic Rework

### PR DESCRIPTION
**Pull request by UnderscoreTud**
### Description
This PR is a rework of my other arithmetic rework :joy: (https://github.com/SkriptLang/Skript/pull/5203), it has the same premise but all the internal stuff is better. This rework also has a different better registration system that makes commutative operations possible, e.g. (type1 - type2) and (type2 - type1) operations, the old one only could support (type1 - type2) due to how it was coded.

closes https://github.com/SkriptLang/Skript/pull/5203

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** https://github.com/SkriptLang/Skript/issues/4253

### CODE REVIEW

File: src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public final class Arithmetics {
+
+	private Arithmetics() {}
+
+	private static final Map<Operator, List<OperationInfo<?, ?, ?>>> registeredOperations = Collections.synchronizedMap(new HashMap<>());
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java:36]: ConcurrentHashMap may be preferred here

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Review by Moderocky — APPROVED** [review]
Looks ok.

File: src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public final class Arithmetics {
+
+	private Arithmetics() {}
+
+	private static final ConcurrentHashMap<Operator, List<OperationInfo<?, ?, ?>>> registeredOperations = new ConcurrentHashMap<>();
+	private static final List<DifferenceInfo<?, ?>> registeredDifferences = Collections.synchronizedList(new ArrayList<>());
+	private static final ConcurrentHashMap<Class<?>, Object> defaultValues = new ConcurrentHashMap<>();
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java:39]:
```suggestion
	private static final Map<Operator, List<OperationInfo<?, ?, ?>>> registeredOperations = new ConcurrentHashMap<>();
	private static final List<DifferenceInfo<?, ?>> registeredDifferences = Collections.synchronizedList(new ArrayList<>());
	private static final Map<Class<?>, Object> defaultValues = new ConcurrentHashMap<>();
```

Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java:39]:
```suggestion
	private static final Map<Operator, List<OperationInfo<?, ?, ?>>> registeredOperations = new ConcurrentHashMap<>();
	private static final List<DifferenceInfo<?, ?>> registeredDifferences = Collections.synchronizedList(new ArrayList<>());
	private static final Map<Class<?>, Object> defaultValues = new ConcurrentHashMap<>();
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/lang/Variable.java [hunk 3]
```diff
@@ -590,53 +592,45 @@ public void change(Event e, @Nullable Object[] delta, ChangeMode mode) throws Un
...
-					if (o == null) {
-						ci = null;
-					} else {
-						Class<?> c = o.getClass();
-						assert c != null;
-						ci = Classes.getSuperClassInfo(c);
-					}
-					Arithmetic a = null;
+					Object object = get(e);
+					Class<?> clazz = object == null ? null : object.getClass();
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/lang/Variable.java:596]: The newly update singleton line doesn't take into consideration `Classes.getSuperClassInfo`

You're removing ClassInfo checking for your isAssignable in Arithmetics.

It should be checking against ClassInfos.

**Review by TheLimeGlass — DISMISSED** [review]

**Comment by TheLimeGlass** [comment]
This pull request is changing so much. I don't believe this should be included in 2.7.

File: src/main/java/ch/njol/skript/lang/Variable.java (hunk 3 above)
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/lang/Variable.java:596]: Why? It works fine

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 4]
```diff
@@ -38,48 +36,46 @@
...
+	private Class<?> returnType;
 	
-	@SuppressWarnings({"unchecked", "null", "unused"})
 	@Override
+	@SuppressWarnings({"unchecked", "unused", "ConstantConditions"})
 	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
 		first = exprs[0];
 		second = exprs[1];
-		final ClassInfo<?> ci;
+		final Class<?> c;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:70]: what does c mean tud
give it a good name 
if you're cold, they're cold

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 5]
```diff
@@ -96,60 +92,60 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
...
+		final Object[] one = (Object[]) Array.newInstance(returnType, 1);
 		
-		// If we're comparing object expressions, such as variables, math is null right now
-		if (relativeType.equals(Object.class)) {
-			ClassInfo<?> info = Classes.getSuperClassInfo(Utils.getSuperType(f.getClass(), s.getClass()));
-			math = info.getMath();
-			if (math == null) { // User did something stupid, just return <none> for them
+		// If we're comparing object expressions, such as variables, difference info is null right now
+		if (returnType.equals(Object.class)) {
+			Class<?> c = Utils.getSuperType(first.getClass(), second.getClass());
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:123]: :c

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		});
+		Arithmetics.registerOperation(Operator.MULTIPLICATION, Number.class, (left, right) -> {
+			if (Utils.isInteger(left, right))
+				return left.longValue() * right.longValue();
+			return left.doubleValue() * right.doubleValue();
+		});
+		Arithmetics.registerOperation(Operator.DIVISION, Number.class, (left, right) -> left.doubleValue() / right.doubleValue());
+		Arithmetics.registerOperation(Operator.EXPONENTIATION, Number.class, (left, right) -> {
+			if (Utils.isInteger(left, right))
+				return (long) Math.pow(left.longValue(), right.longValue());
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:50]: This causes `3^-1` to return `0`.
You can only return a long if the right value is also positive.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
This changes behavior when something like a string variable is used in arithmetic.
Before, any non-number is treated as 0 from the get-go. Now, the first operation involving the string will return null, and any uses of that null from then on will be treated as 0.
```
{_x} = "a"
before: {_x} / 2 = 0, 2  ^ {_x} + 1 = 2
now: {_x}  / 2 = <none>, 2 ^ {_x} + 1 = 1
```

I will test vectors and timespans at a later date but i need to sleep.

Would also like to see tests!

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+			Vector rightVector = new Vector(number, number, number);
+			return left.clone().divide(rightVector);
+		}, (left, right) -> {
+			double number = left.doubleValue();
+			Vector leftVector = new Vector(number, number, number);
+			return leftVector.divide(right);
+		});
+		Arithmetics.registerDifference(Vector.class,
+			(left, right) -> new Vector(Math.abs(left.getX() - right.getX()), Math.abs(left.getY() - right.getY()), Math.abs(left.getZ() - right.getZ())));
+		Arithmetics.registerDefaultValue(Vector.class, Vector::new);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:84]: This should be up with the other vector-vector stuff imo

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+		// Timespan - Timespan
+		Arithmetics.registerOperation(Operator.ADDITION, Timespan.class, (left, right) -> new Timespan(left.getMilliSeconds() + right.getMilliSeconds()));
+		Arithmetics.registerOperation(Operator.SUBTRACTION, Timespan.class, (left, right) -> new Timespan(Math.max(0, left.getMilliSeconds() - right.getMilliSeconds())));
+		Arithmetics.registerDifference(Timespan.class, (left, right) -> new Timespan(Math.abs(left.getMilliSeconds() - right.getMilliSeconds())));
+		Arithmetics.registerDefaultValue(Timespan.class, Timespan::new);
+
+		// Timespan - Number
+		Arithmetics.registerOperation(Operator.MULTIPLICATION, Timespan.class, Number.class, (left, right) -> new Timespan(left.getMilliSeconds() * right.longValue()));
+		Arithmetics.registerOperation(Operator.DIVISION, Timespan.class, Number.class, (left, right) -> new Timespan(left.getMilliSeconds() / right.longValue()));
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:94]: Dividing by 0, NaN,  or by a non-number results in a stacktrace:
```
[18:38:31 ERROR]: #!#! Stack trace:
[18:38:31 ERROR]: #!#! java.lang.ArithmeticException: / by zero
[18:38:31 ERROR]: #!#!     at Skript.jar//ch.njol.skript.classes.data.DefaultOperations.lambda$static$20(DefaultOperations.java:94)
[18:38:31 ERROR]: #!#!     at Skript.jar//ch.njol.skript.expressions.arithmetic.ArithmeticChain.calculate(ArithmeticChain.java:96)
[18:38:31 ERROR]: #!#!     at Skript.jar//ch.njol.skript.expressions.arithmetic.ArithmeticChain.get(ArithmeticChain.java:82)
```

Negative numbers (or infinity) when multiplying and dividing result in illegal argument stacktrace:
```
[18:41:30 ERROR]: #!#! Stack trace:
[18:41:30 ERROR]: #!#! java.lang.IllegalArgumentException: millis must be >= 0
[18:41:30 ERROR]: #!#!     at Skript.jar//ch.njol.skript.util.Timespan.<init>(Timespan.java:142)
[18:41:30 ERROR]: #!#!     at Skript.jar//ch.njol.skript.classes.data.DefaultOperations.lambda$static$20(DefaultOperations.java:94)
[18:41:30 ERROR]: #!#!     at Skript.jar//ch.njol.skript.expressions.arithmetic.ArithmeticChain.calculate(ArithmeticChain.java:96)
```

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		Arithmetics.registerDefaultValue(Vector.class, Vector::new);
+
+		// Timespan - Timespan
+		Arithmetics.registerOperation(Operator.ADDITION, Timespan.class, (left, right) -> new Timespan(left.getMilliSeconds() + right.getMilliSeconds()));
+		Arithmetics.registerOperation(Operator.SUBTRACTION, Timespan.class, (left, right) -> new Timespan(Math.max(0, left.getMilliSeconds() - right.getMilliSeconds())));
+		Arithmetics.registerDifference(Timespan.class, (left, right) -> new Timespan(Math.abs(left.getMilliSeconds() - right.getMilliSeconds())));
+		Arithmetics.registerDefaultValue(Timespan.class, Timespan::new);
+
+		// Timespan - Number
+		Arithmetics.registerOperation(Operator.MULTIPLICATION, Timespan.class, Number.class, (left, right) -> new Timespan(left.getMilliSeconds() * right.longValue()));
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:93]: This one should be commutative, `2 * 2 seconds` should equal `4 seconds`, not `0`.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
The non-number behavior is a lot better but still not quite right. It does work as expected for null values, but not when the variable is a string.
```
    set {_y} to "2"
    broadcast "%{_y} + 1 + 1% = 2" # broadcasts 1
    broadcast "%1 + {_y} + 1% = 2" # broadcasts 2
```
The same thing occurs for vectors. It seems to depend on what the left operators is:
```
<none> = <0, 0, 0>
```
I've attached the full test file to this report.
[arithmetic-test.txt](https://github.com/SkriptLang/Skript/files/13465799/arithmetic-test.txt)

***

So far I've been testing assuming the null values should be the 0 for whatever they are, but for the following example I'm not sure what the answer should be:
```
{_vector} + {_number}
``` 
Should it be ``{_vector} + vector(0,0,0)``? or ``0 + {_number}``? I'm not sure which would be preferable, and it gets worse with chained oop stuff. Perhaps any invalid operation should just return NaN or none and leave it as that. It'd be breaking, but it'd be consistent.

This is the same with like `{_timespan} + 1`. Reading the code w/out context makes it seem like it should be a number, but it'll actually be a timespan.

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+			long scalar = right.longValue();
+			if (scalar <= 0)
+				return null;
+			return new Timespan(left.getMilliSeconds() / scalar);
+		}, (left, right) -> {
+			long scalar = left.longValue();
+			if (scalar < 0)
+				return null;
+			return new Timespan(scalar / right.getMilliSeconds());
+		});
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:115]: I can see why you made division commutative too, but I don't think it should be.
1 / (1 second) should really be 1 second^-1, not 1 second, so i think number / timespan should be invalid.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
I can't find anything else
nice work tud
[arithmetic-test.txt](https://github.com/SkriptLang/Skript/files/13520378/arithmetic-test.txt)

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java (hunk 10 above)
Comment (UnderscoreTud) [reply src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:115]: > 1 / (1 second) should really be 1 second^-1

So you're saying that I should add an exponentiation operation for timespan - number?

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:115]: no no
second^-1 is notation for "per second"
i'm just saying you shouldn't be able to divide a number by a timespan

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/org/skriptlang/skript/lang/arithmetic/OperationInfo.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+			return null;
+		if (!Converters.converterExists(fromLeft, left) || !Converters.converterExists(fromRight, right) || !Converters.converterExists(returnType, toReturnType))
+			return null;
+		return new OperationInfo<>(fromLeft, fromRight, toReturnType, (left, right) -> {
+			L convertedLeft = Converters.convert(left, this.left);
+			R convertedRight = Converters.convert(right, this.right);
+            if (convertedLeft == null || convertedRight == null)
+				return null;
+            T result = operation.calculate(convertedLeft, convertedRight);
+			return Converters.convert(result, toReturnType);
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/arithmetic/OperationInfo.java:74]:
```suggestion
		if (convertedLeft == null || convertedRight == null)
			return null;
		T result = operation.calculate(convertedLeft, convertedRight);
			return Converters.convert(result, toReturnType);
```

File: src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	public static <T> void registerDifference(Class<T> type, Operation<T, T, T> operation) {
+		registerDifference(type, type, operation);
+	}
+
+	public static <T, R> void registerDifference(Class<T> type, Class<R> returnType, Operation<T, T, R> operation) {
+		Skript.checkAcceptRegistrations();
+		if (exactDifferenceExists(type))
+			throw new IllegalArgumentException("There's already a difference registered for type '" + type + "'");
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/arithmetic/Arithmetics.java:184]: SkriptAPIException? :)

File: src/main/java/org/skriptlang/skript/lang/arithmetic/OperationInfo.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+		});
+	}
+
+	@Override
+	public String toString() {
+		return String.join(", ", "OperationInfo[left=" + left,
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/arithmetic/OperationInfo.java:80]: guava has something like MoreObjects I think you can use here

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Review by APickledWalrus — APPROVED** [review]
It's ready for primetime

File: src/test/skript/tests/syntaxes/expressions/ExprArithmetic.sk [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+	return false if component_equals((z of {_vec}), {_z}) is false
+	return true
+
+local function component_equals(a: number, b: number) :: boolean:
+	if:
+		isNaN({_a}) is true
+		isNaN({_b}) is true
+	then:
+		return true
+	return true if {_a} is {_b}, else false
```
Comment (APickledWalrus) [inline src/test/skript/tests/syntaxes/expressions/ExprArithmetic.sk:257]:
```suggestion
	return true if {_a} is {_b}, else false

```

File: src/main/java/ch/njol/skript/expressions/ExprDifference.java [hunk 15]
```diff
@@ -156,34 +151,27 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
...
-		Arithmetic math = this.math;
-		Class<?> relativeType = this.relativeType;
-
-		if (relativeType == Object.class) { // Try to determine now that actual types are known
-			ClassInfo<?> info = Classes.getSuperClassInfo(Utils.getSuperType(first.getClass(), second.getClass()));
-			math = info.getMath();
-			if (math == null) { // User did something stupid, just return <none> for them
+		if (differenceInfo == null) { // Try to determine now that actual types are known
+			Class<?> superType = Utils.getSuperType(first.getClass(), second.getClass());
+			differenceInfo = Arithmetics.getDifferenceInfo(superType);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprDifference.java:164]:
```suggestion
		DifferenceInfo differenceInfo = this.differenceInfo;
		if (differenceInfo == null) { // Try to determine now that actual types are known
			Class<?> superType = Utils.getSuperType(first.getClass(), second.getClass());
			differenceInfo = Arithmetics.getDifferenceInfo(superType);
```
We do not want to modify the actual field at runtime

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprArithmetic.sk [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+test "date arithmetic":
+    set {_d1} to now
+    set {_d2} to 1 day ago
+    assert ({_d1} + {_d2}) is not set with ""
+
+    assert ({_d1} + 1 day) is (1 day from now) with "now + 1 day is not 1 day from now"
+    assert ({_d1} - 1 day) is (1 day ago) with "now - 1 day is not 1 day ago"
+    assert ({_d1} + 1 week) is (1 week from now) with "now + 1 week is not 1 week from now"
+    assert ({_d1} - 1 week) is (1 week ago) with "now - 1 week is not 1 week ago"
```
Comment (APickledWalrus) [inline src/test/skript/tests/syntaxes/expressions/ExprArithmetic.sk:242]:
```suggestion
    assert ({_d1} + 1 day) is (1 day from {_d1}) with "now + 1 day is not 1 day from now"
    assert ({_d1} - 1 day) is (1 day ago) with "now - 1 day is not 1 day ago"
    assert ({_d1} + 1 week) is (1 week from {_d1}) with "now + 1 week is not 1 week from now"
    assert ({_d1} - 1 week) is (1 week ago) with "now - 1 week is not 1 week ago"
```

**Review by APickledWalrus — APPROVED** [review]

**Pull request merged**